core/go/internal/publictxmgr/transaction_manager.go

```diff
@@ -352,6 +391,103 @@ func (ptm *pubTxManager) WriteNewTransactions(ctx context.Context, dbTX persiste
+	log.L(ctx).Debugf("WriteReceivedTransactions transactions: %+v", transactions)
+	persistedTransactions := make([]*DBPublicTxn, len(transactions))
+	for i, txi := range transactions {
+		persistedTransactions[i] = &DBPublicTxn{
+			From:            *txi.From, // safe because validated in ValidateTransaction
+			To:              txi.To,
+			Gas:             txi.Gas.Uint64(),
+			Value:           txi.Value,
+			Data:            txi.Data,
+			FixedGasPricing: pldtypes.JSONString(txi.PublicTxGasPricing),
+		}
+	}
+	// All the nonce processing to this point should have ensured we do not have a conflict on nonces.
+	// It is the caller's responsibility to ensure we do not have a conflict on transaction+resubmit_idx.
+	if len(persistedTransactions) > 0 {
+		log.L(ctx).Debugf("WriteReceivedTransactions persisted transactions: %d", len(persistedTransactions))
+		err = dbTX.DB().
+			WithContext(ctx).
+			Table("public_txns").
+			Clauses(clause.Returning{Columns: []clause.Column{{Name: "pub_txn_id"}}}).
+			Create(persistedTransactions).
+			Error
+	}
+	if err == nil {
+		publicTxBindings := make([]*DBPublicTxnBinding, 0, len(transactions))
+		for i, txi := range transactions {
+			pubTxnID := persistedTransactions[i].PublicTxnID
+			log.L(ctx).Debugf("WriteReceivedTransactions public txn binding ID: %+v", pubTxnID)
+			for _, bnd := range txi.Bindings {
+				publicTxBindings = append(publicTxBindings, &DBPublicTxnBinding{
+					Transaction:     bnd.Transaction,
+					TransactionType: bnd.TransactionType,
+					Sender:          bnd.TransactionSender,
+					ContractAddress: bnd.TransactionContractAddress,
+					PublicTxnID:     pubTxnID,
+				})
+			}
+		}
+		if len(publicTxBindings) > 0 {
+			log.L(ctx).Debugf("WriteReceivedTransactions public txn bindings: %d", len(persistedTransactions))
+			err = dbTX.DB().
+				WithContext(ctx).
+				Table("public_txn_bindings").
+				Create(publicTxBindings).
+				Error
+		}
+	}
+
+	return pubTxns, err
+}
+
+func (ptm *pubTxManager) WriteReceivedPublicTransactionSubmissions(ctx context.Context, dbTX persistence.DBTX, submissions []*pldapi.PublicTxToDistribute) (pubTxns []*pldapi.PublicTx, err error) {
+	log.L(ctx).Debugf("WriteReceivedTransactions transactions: %+v", submissions)
+	persistedSubmissions := make([]*DBPubTxnSubmission, 0, len(submissions))
+	for _, submission := range submissions {
+		for _, binding := range submission.Bindings {
+			dbSubmission := &DBPubTxnSubmission{
+				TransactionHash: *submission.TransactionHash,
+				GasPricing:      submission.GasPricing.Bytes(),
+				PrivateTXID:     binding.Transaction,
+			}
+			persistedSubmissions = append(persistedSubmissions, dbSubmission)
+
+			err = dbTX.DB().
+				Table("public_txn_bindings").
+				Select(`"pub_txn_id"`).
+				Where(`"transaction" IN (?)`, binding.Transaction).
+				Find(&dbSubmission.PublicTxnID).
+				Error
```

Comment:
Seems inefficient to do a single DB query for each item, vs. an aggregated query (there's already an `IN` clause here)